test/metabase/lib/swap_test.cljc

```diff
@@ -0,0 +1,123 @@
+            (and ;; Correctly rearranged the clauses.
+                 (= (for [clause (clause-fn query)]
+                      (cond
+                        (= source-clause clause) target-clause
+                        (= target-clause clause) source-clause
+                        :else                    clause))
+                    (clause-fn swapped))
+                 ;; And didn't change anything else.
+                 (= (m/dissoc-in query   [:stages 0 clause-key])
+                    (m/dissoc-in swapped [:stages 0 clause-key]))))))))
+
+(defspec swap-clauses-on-aggregations-test-permutations
+  (swapped-clauses-prop
+    lib/aggregations :aggregation
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/aggregate (lib/count))
+        (lib/aggregate (lib/sum (meta/field-metadata :orders :subtotal)))
+        (lib/aggregate (lib/avg (meta/field-metadata :orders :discount)))
+        (lib/aggregate (lib/max (meta/field-metadata :orders :tax)))
+        (lib/aggregate (lib/min (meta/field-metadata :orders :total))))))
+
+(defspec swap-clauses-on-breakouts-test-permutations
+  (swapped-clauses-prop
+    lib/breakouts :breakout
+    (as-> (lib/query meta/metadata-provider (meta/table-metadata :orders)) $q
+      (lib/breakout $q (meta/field-metadata :products :category))
+      (lib/breakout $q (meta/field-metadata :people :source))
+      (lib/breakout $q (lib/with-binning
+                         (meta/field-metadata :orders :subtotal)
+                         (second (lib/available-binning-strategies $q (meta/field-metadata :orders :subtotal)))))
+      (lib/breakout $q (lib/with-temporal-bucket (meta/field-metadata :orders :created-at) :month)))))
+
+(defspec swap-clauses-on-filters-test-permutations
+  (swapped-clauses-prop
+    lib/filters :filters
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/filter (lib/= (meta/field-metadata :products :category) "Doohickey"))
+        (lib/filter (lib/< (meta/field-metadata :products :created-at) "2024-01-01T00:00:00"))
+        (lib/filter (lib/is-null (meta/field-metadata :orders :discount)))
+        (lib/filter (lib/!= (meta/field-metadata :people :source) "Facebook")))))
+
+(defspec swap-clauses-on-expressions-test-permutations
+  (swapped-clauses-prop
+    lib/expressions :expressions
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/expression "clean-discount" (lib/coalesce (meta/field-metadata :orders :discount) 0))
+        (lib/expression "discount fraction" (lib// (meta/field-metadata :orders :discount)
+                                                   (meta/field-metadata :orders :subtotal)))
+        (lib/expression "order month" (lib/get-month (meta/field-metadata :orders :created-at)))
+        (lib/expression "signup year" (lib/get-year  (meta/field-metadata :people :created-at))))))
```

Comment:
What about expressions that depend on a previous expression?

Reply:
Expressions are not ordered (semantically) AFAIK. They will end up siblings in a `SELECT ... FROM` eventually, and those can be in any order.

Actually today I wrote a complicated expression, then split it up into two for ease of understanding. I edited the first/left one to reference the second and it was all good.

Reply:
Thanks, I was surprised this work, but it does because the qp runs substitution on expression references so `[two] + 2`  turns into native `(1 + 1) + 2`

```
(let [mp (lib.metadata.jvm/application-database-metadata-provider (mt/id))
      query (-> (lib/query mp (lib.metadata/table mp (mt/id :products)))
                       (lib/expression "two" (lib/+ 1 1))
                       (as-> $q (lib/expression $q "four" (lib/+ (lib/expression-ref $q "two") 2))))
      query (update-in query [:stages 0 :expressions] reverse)]
 (qp/process-query query))
 ```